Split implementation to a pure app

ulrikstrid opened the pull request:
This is a pretty early PR that just moves the code around a bit to not depend on cohttp.
Tests seems to pass on my machine at least.

Closes #39

talex5 commented on prometheus-app-pure.opam:
```diff
@@ -0,0 +1,34 @@
+opam-version: "2.0"
+synopsis: "Client library for Prometheus monitoring"
+description: """\
+Applications can enable metric reporting using the `prometheus-app` opam package.
+This depends on cohttp and can serve the metrics collected above over HTTP.
+
+The `prometheus-app.unix` ocamlfind library provides the `Prometheus_unix` module,
+which includes a cmdliner option and pre-configured web-server.
+See the `examples/example.ml` program for an example, which can be run as:
+
+Unikernels can use `Prometheus_app` instead of `Prometheus_unix` to avoid the `Unix` dependency."""
```
This needs updating. It's describing the app package.

talex5 reviewed (commented):
Looks reasonable. I'm not keen on the name `prometheus-app-pure` though (especially as it has the side-effect of registering collectors!). Maybe `prometheus-reporter`?

ulrikstrid commented:
Sorry for the super late update here @talex5. I have now rebased, changed the name to `prometheus-reporter` and changed the description in the opam file.

talex5 commented on prometheus-reporter.opam:
```diff
@@ -0,0 +1,33 @@
+
+The `prometheus-reporter.unix` ocamlfind library provides the `Prometheus_reporter_unix` module,
+which includes a cmdliner option.
+See the `examples/example.ml` program for an example, which can be run as:
```
Missing text here?

talex5 commented on reporter/prometheus_reporter.mli:
```diff
@@ -0,0 +1,18 @@
+(** Report metrics for Prometheus.
+    See: https://prometheus.io/
```
This was `See: {{:https://prometheus.io/}https://prometheus.io/}` previously.

talex5 commented on reporter/prometheus_reporter_unix.mli:
```diff
@@ -0,0 +1,55 @@
+(** Report metrics for Prometheus.
+    See: https://prometheus.io/
```
Should also be a link.

talex5 commented on reporter/prometheus_reporter_unix.mli:
```diff
@@ -0,0 +1,55 @@
+    - This module automatically initialises itself and registers some standard collectors relating to
+      GC statistics, as recommended by Prometheus.
+
+    - This extends [Prometheus_reporter] with support for cmdliner option parsing, a server pre-configured
+      for Unix, and a start-time metric that uses [Unix.gettimeofday].
+ *)
+
+type config = int option
+
+val opts : config Cmdliner.Term.t
+(** [opts] is the extra command-line options to offer Prometheus
+    monitoring. *)
```
It might make more sense to keep the cmdliner stuff for cohttp in app. That would also avoid exposing the type (which is likely to change - see #51).

talex5 commented on app/prometheus_app.ml:
```diff
@@ -1,147 +1,4 @@
-    process_cpu_seconds_total;
-  ]
-end
+include Prometheus_reporter
```
We should also be able to include it in the mli, to avoid duplication.

smorimoto commented:
I would love to see a combination of Prometheus and Dream!

avsm commented:
Now fixed in #71 based on this PR. Thanks for the contribution!
